design/go-api.md

```diff
@@ -0,0 +1,72 @@
+# Replace the current typescript kurl.sh api for creating installers with a go api using gin
+ 
+This proposal outlines the benefits and challenges of replacing the current typescript api with a more simple golang api.
+
+## Goals
+
+- Significantly reduce code complexety around making changes to the installer type
+- Provide type safety and validation for POSTed installers
+
+## Non Goals
+
+- Create an API for kurlkinds dsecriptions. This probably best suited in the kurl.sh site repo as it is not core functionality of the api.
+
+## Background
+
+Previously, kURL has used the backend typescript api to take a yaml config and make an install script with options and bash variables set for the install.
+With the recent changes to kURL such as the addition of the CRD, the current install script uses go binaries to parse and traverse the yaml tree to set options.
+
+We can greatly reduce code complexity from the no longer used parts of the typescript api and gain added type safety by moving this api to Go.
+
+## High-Level Design
+
+Endpoints
+
+postInstaller POST /installer
+putInstaller PUT /installer/{name}
+getInstallScript GET /{installerID}
+getJoinScript GET /{installerID}/join.sh
+getUpgradeScript GET /{installerID}/upgrade.sh
+getInstallerYaml GET /installer/{installerID}
+validateInstaller POST /installer/validate
+getBundle GET/bundle/{installerID}
+
+## Detailed Design
+
+postInstaller Endpoint:
+The function will take a yaml Installer and attempt to unmarshall it into an Installer CRD.
+If the unmarshalling fails, an error will be retuned. If the marshalling succeeds the version of an addon will be checked against existing versions.
+If the version is not found an error will be returned.
```

Comment:
There are a couple other validations besides version. The podCidrRange and serviceCidrRange are checked by regex. Will those be ported over? We should add more validations like that in the future.

Reply:
Right I see those, added if there are any more we can them add too